Add ADMM + parameter reordering FISTA

jeverink opened the pull request:
By request, sub PR from https://github.com/CUQI-DTU/CUQIpy/pull/431

Added ADMM solver

closes #562

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
     
+    Parameters
+    ----------
+    A : ndarray or callable f(x,*args).
```
I think "f" is a bit misleading here. Maybe just do this
```suggestion
    A : ndarray or callable with arguments (x, *args).
```

jeverink replied:
I think I copied this line directly form the CGLS or FISTA documentation, so such update would have to be done to all.

I agree that removing the explicit mention of f could help. Perhaps we could also explicitly mention what the arguments are, i.e., the flag parameter and what is expected of this. Do you have any suggestions of how this can be done properly in the documentation?

nabriis replied:
Something like
```suggestion
    A : ndarray or callable
        Represents a matrix or a function that performs matrix-vector multiplications. 
        When A is a callable, it accepts arguments (x, flag) where:
        - flag=1 indicates multiplication of A with vector x, that is A @ x.
        - flag=2 indicates multiplication of the transpose of A with vector x, that is  A.T @ x.
```

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+    b : ndarray.
+    penalties : List of tuples (callable proximal operator of f_i, linear operator L_i).
+    x0 : ndarray. Initial guess.
+    tradeoff : Trade-off between linear least squares and regularization term in the solver iterates.
```
Is "tradeoff" the term used in the Boyd reference? If not can you add a a bit more explanation. If it is a term used, can you add to see the reference paper?

jeverink replied:
The original name used for this in the reference is "penalty parameter", however, as I refer to the list of regularization/constraints as penalties, I thought this would be confusing.  Alternatively, the character "ρ" is used in the equations of the reference, hence why it is called rho inside the class. I thought that "tradeoff" would better describe the interpretation of the variable.

Any suggestions on how/what to modify?

nabriis replied:
```suggestion
    tradeoff : Trade-off between linear least squares and regularization term in the solver iterates. Denoted as "rho" in [1].
```
Then add [1] as your reference.

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+    -----------
+    .. code-block:: python
+    
+        from cuqi.solver import ADMM,  ProximalL1, ProjectNonnegative
```
```suggestion
        from cuqi.solver import ADMM, ProximalL1, ProjectNonnegative
```

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -678,6 +794,22 @@ def ProjectBox(x, lower = None, upper = None):
+    b : array_like.
+    """  
+
+    ax_b = np.inner(a,x)-b
```
```suggestion
    ax_b = np.inner(a,x) - b
```

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+
+    """  
+
+    def __init__(self, A, b, penalties, x0, tradeoff, maxit = 100, adaptive = True):
```
We usually really like to have defaults on as many parameters as possible in CUQIpy. Is it possible to use some defaults for "x0" and "tradeoff"

jeverink replied:
I've added it for "tradeoff". For initial guess "x0", this should probably be added then for all solvers.

nabriis replied:
Aight. Thanks. I added this issue #556

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+        self.maxit = maxit
+        self.adaptive = adaptive
+
+        self.penalties = penalties
```
I suggest making many of these attributes private (e.g. _penalties), so it is clear they cannot be modified after initialization.

jeverink replied:
Perhaps update this as part of a separate refactor of all solvers, as this is made to align with the others.

nabriis replied:
#557

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+        self.u_cur = [np.zeros(l) for l in dual_len]
+        self.n = penalties[0][1].shape[1]
+        
+        self.rho = tradeoff
```
Here we are mixing rho and tradeoff. Since we actually used A and b in the input parameters, perhap "rho" can also be used there. Please make sure it is included in the equation also in the docstring.

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+        # Iterating
+        for i in range(self.maxit):
+            # Preprocessing
+            if callable(self.A):
+                def big_matrix(x, flag):
+                    if flag == 1:
+                        out1 = np.sqrt(1/self.rho)*self.A(x, 1)
+                        out2 = [penalty[1]@x for penalty in self.penalties]
+                        out  = np.hstack([out1] + out2)
+                    elif flag == 2:
+                        idx_start = len(x)
+                        idx_end = len(x)
+                        out1 = np.zeros(self.n)
+                        for _, t in reversed(self.penalties):
+                            idx_start -= t.shape[0]
+                            out1 += t.T@x[idx_start:idx_end]
+                            idx_end = idx_start
+                        out2 = np.sqrt(1/self.rho)*self.A(x[:idx_end], 2)
+                        out  = out1 + out2     
+                    return out
+            else:
+                big_matrix = np.vstack([np.sqrt(1/self.rho)*self.A] + [penalty[1] for penalty in self.penalties])
```
I suggest refactoring this whole part into a method

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+        self.p = len(self.penalties)
+
+    def solve(self):
+        y_new = self.p*[0]
```
Could you add a docstring here on the solve method.

Could you also add an explanation of the general approach for solving. This could include a bit of details on the penalties and the "big_matrix" and "big_vector". I think added a docstring and renaming those to some variables defined in the docstring could be helpful for understanding.

jeverink replied:
I have added the reference here and changed some of the variable names to align with the reference.

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+            # Main update (Least Squares)
+            big_vector = np.hstack([np.sqrt(1/self.rho)*self.b] + [self.y_cur[i] - self.u_cur[i] for i in range(self.p)])
+
+            solver = CGLS(big_matrix, big_vector, self.x_cur, 10)
```
Do we want this "inner_max_it" to be configurable?

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+            res_dual = 0.0
+            for j, penalty in enumerate(self.penalties):
+                res_dual += LA.norm(penalty[1].T@(y_new[j] - self.y_cur[j]))**2
+            #res_dual *= self.rho*self.rho
```
There seems to be some commented code here.

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+                res_dual += LA.norm(penalty[1].T@(y_new[j] - self.y_cur[j]))**2
+            #res_dual *= self.rho*self.rho
+
+            if self.adaptive:
```
Is this adaptive scheme described in Boyd reference? If so can you add a comment on where it comes from?

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,124 @@ def solve(self):
+
+    def __init__(self, A, b, penalties, x0, tradeoff, maxit = 100, adaptive = True):
+
+        self.A = A
```
Could we add some input checking "guarding" statements here that check the values of the parameters are as expected and throw errors if not?

nabriis replied:
#558

nabriis commented on tests/test_solver.py:
```diff
@@ -54,8 +54,27 @@ def test_FISTA():
\ No newline at end of file
+    assert np.allclose(sol, ref_sol, atol=1e-4)
+
+def test_ADMM():
```
This unit test does indeed seem to reach most code paths in the ADMM class. It seems that we are not testing if A was passed as a callable function. Could we add a test for that also?

nabriis reviewed (changes requested):
Thank you for splitting the PR into these smaller parts @jeverink!

I appreciate you adding this ADMM solver, including the adaptive scheme.

I know the current solver module is a bit lackluster with respect to code quality, but I was thinking that this PR can start moving the overall state of the solver module in a better direction. Therefore I added quite a bit of comments and suggestions.

Do not hesitate to reach out if you have questions on the comments.

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,148 @@ def solve(self):
+
+        # Iterating
+        for i in range(self.maxit):
+            big_matrix, big_vector = self._iteration_pre_processing()
```
It seems that the new method here is storing self._big_matrix as an attribute in the class instance, but also working with a local variable "big_matrix" here. Perhaps it might be best to use just self._big_matrix to avoid potential confusion in which variable is being used. If needed you can also store self._big_vector and just use those. Then the method would be:
```python
self._iteration_pre_processing()
```

nabriis commented on tests/test_solver.py:
```diff
@@ -54,8 +54,56 @@ def test_FISTA():
+    # Compare
+    assert np.allclose(sol, ref_sol, atol=1e-4)
+
+test_ADMM_function_form()
```
There seem to be a duplicate method here? Not sure if that is the cause for the unit tests failing. It looked like they began failing when the test "test_ADMM_function_form" was added.

nabriis reviewed (changes requested):
Thank you @jeverink. I agree with your comments on code style etc. I created 3 issues. I only have a few comments left related to the code. I think we are nearly there! The refactor of the pre-processing was really helpful for readability. Thank you for adding the docstrings also!

jeverink commented:
Thanks for the help @nabriis,
I have now fixed the test (I accidentally copy-pasted that one line that broke it all), and made some more final small improvements.

nabriis reviewed (approved):
Thanks @jeverink. LGTM!

amal-ghamdi commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,154 @@ def solve(self):
+
+            # Adaptive approach based on [1], Subsection 3.4.1
+            if self.adaptive:
+                if res_dual > 1e2*res_primal:
```
is the number 1e2 fixed? if it is useful to configure by the user, it can be an input to the initializer

jeverink replied:
This number and the numbers in the lines below are describes as "typical choices", with a relatively easy interpretation.- I have updated the documentation to mention this.

amal-ghamdi reviewed (commented):
Thank you @jeverink for your contribution, great work!

I made a few comments, please feel free to address them as you see fit and let me know if any of my comments are unclear.

jeverink commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,154 @@ def solve(self):
+        self.inner_max_it = inner_max_it
+        self.adaptive = adaptive
+
+        self.penalties = penalties
```
I propose we let this fall under #557

amal-ghamdi replied:
👍

jeverink commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,154 @@ def solve(self):
+
+class ADMM(object):
+    """Alternating Direction Method of Multipliers for solving regularized linear least squares problems of the form:
+    Minimize ||Ax-b||^2 + sum_i f_i(L_i x).
```
I have added a comment regarding the range of summation and there is a corrected example at the end of this docstring. If you propose to add a bit more information, could you give an example of what you think could be helpful.

jeverink commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,154 @@ def solve(self):
+    ----------
+    A : ndarray or callable
+        Represents a matrix or a function that performs matrix-vector multiplications. 
+        When A is a callable, it accepts arguments (x, flag) where:
+        - flag=1 indicates multiplication of A with vector x, that is A @ x.
+        - flag=2 indicates multiplication of the transpose of A with vector x, that is  A.T @ x.
```
#561

jeverink commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,154 @@ def solve(self):
+        A = rng.standard_normal((m, n))
+        b = rng.standard_normal(m)
+        x0 = np.zeros(n)
+        admm = ADMM(A, b, x0, penalties = [(ProximalL1, np.eye(n)), (lambda z, _ : ProjectNonnegative(z))], tradeoff = 10)
```
Fixed

jeverink commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,154 @@ def solve(self):
+    b : ndarray.
+    penalties : List of tuples (callable proximal operator of f_i, linear operator L_i).
+    x0 : ndarray. Initial guess.
+    tradeoff : Trade-off between linear least squares and regularization term in the solver iterates. Denoted as "rho" in [1].
```
Regarding the first two bullet points: I have changed the names.
Regarding the third bullet point: Describing it as a trade-off between the misfit and regulatization is the best I have come up with. It is not part of the original optimization problem and adding it to a formula requires formulating an augmented Lagrangrian, which I believe is to much detail to put into the docstring and the full details can be found in the reference.

amal-ghamdi replied:
sounds good. If it makes sense you can add a general statement like "Larger penalty_parameter gives larger weight to the penalty terms." 

Just to give an idea if this parameter gives more weight to the misfit or regularization

jeverink commented on cuqi/solver/_solver.py:
```diff
@@ -650,8 +650,154 @@ def solve(self):
+        
+            # Regularization update
+            for j, penalty in enumerate(self.penalties):
+                z_new[j] = penalty[0](penalty[1]@x_new + self.u_cur[j], 1.0/self.rho)
```
I have added additional line in the documentation regarding the input of the proximal operators

amal-ghamdi replied:
👍

jeverink commented:
Thanks for the feedback @amal-ghamdi,
I have now updated the documentation and changed some of the variable names to be more inline with the reference article.

amal-ghamdi commented on cuqi/solver/_solver.py:
```diff
@@ -653,7 +653,8 @@ def solve(self):
     """Alternating Direction Method of Multipliers for solving regularized linear least squares problems of the form:
-    Minimize ||Ax-b||^2 + sum_i f_i(L_i x).
+    Minimize ||Ax-b||^2 + sum_i f_i(L_i x),
+    where the sum ranges from 1 to an arbitrary n.
```
```suggestion
    where the sum ranges from 1 to an arbitrary n. See definition of the parameter `penalty_terms` below for more details about f_i and L_i
```

amal-ghamdi reviewed (approved):
Thank you @jeverink! it looks great. I added one line suggestion in the description of the optimization functional and one comment about the documentation of the penalty_parameter for your consideration.
I am approving the PR.
